build: Require C++17

davidgumberg opened the pull request:
`std::is_standard_layout_v` (used in `util/no_destructor.h` is not available until C++-17. This breaks building with some compilers.

Tested with clang 1f1c725b68b9d72b5cd3fc95c57ce4244bc85c8e (https://github.com/llvm/llvm-project/commit/1f1c725b68b9d72b5cd3fc95c57ce4244bc85c8e)

https://github.com/google/leveldb/pull/1246
https://en.cppreference.com/w/cpp/types/is_standard_layout.html

I ran into this when trying to build with a master `clang` to test #54.

pwnall committed:
Allow different C/C++ standards when this is used as a subproject.

Inspired by https://github.com/google/snappy/pull/85

PiperOrigin-RevId: 279649967

fanquake commented:
If we are going to change this, I think you should just change it to C++20.

l0rinc commented:
An alternative would be cherry-picking https://github.com/google/leveldb/pull/1251

fanquake commented:
We aren't trying to retain C++11 compatibility?

fanquake commented:
You could also update the CMake minimum required to `3.22` (just matching the Core minimum required).

maflcko commented:
If this started happening recently in clang, i think we can wait for google to bump theirs (last bump: https://chromium.googlesource.com/chromium/src/tools/clang/+/4e2e7b926a033fc85695da101404f79b7ce2d7a7) and then wait for them to pull in one of the fixes?

davidgumberg commented:
> An alternative would be cherry-picking [google#1251](https://github.com/google/leveldb/pull/1251)

I considered this fix, but it seems preferable to me to maintain a build flag patch over upstream rather than a code patch.

> If this started happening recently in clang, i think we can wait for google to bump theirs (last bump: [chromium.googlesource.com/chromium/src/tools/clang/+/4e2e7b926a033fc85695da101404f79b7ce2d7a7](https://chromium.googlesource.com/chromium/src/tools/clang/+/4e2e7b926a033fc85695da101404f79b7ce2d7a7)) and then wait for them to pull in one of the fixes?

This is already an issue with clang 20, e.g. my distro's included version of clang `20.1.6` fails to compile with this error.

maflcko commented:
> This is already an issue with clang 20, e.g. my distro's included version of clang `20.1.6` fails to compile with this error.

The downstream CI is using clang 20 ( https://github.com/bitcoin/bitcoin/blob/5757de4ddd37f9321ee6b338b40888fd3561fc00/ci/test/00_setup_env_native_fuzz.sh#L11, https://github.com/bitcoin/bitcoin/blob/5757de4ddd37f9321ee6b338b40888fd3561fc00/ci/test/01_base_install.sh#L53 , ...) is there a reason why it passes?

fanquake commented:
We don't use the CMakeFiles from this repo; we just compile the objects we want directly, using our own options.

maflcko commented:
Ah ok, so this is a dev-only test-only fix for this repo.

sedited commented on CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@
 # found in the LICENSE file. See the AUTHORS file for names of contributors.
 
-cmake_minimum_required(VERSION 3.9)
+cmake_minimum_required(VERSION 3.22)
```
What's the harm in leaving this as is?

hebasto replied:
Because of this warning:
```
CMake Deprecation Warning at CMakeLists.txt:5 (cmake_minimum_required):
  Compatibility with CMake < 3.10 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value.  Or, use the <min>...<max> syntax
  to tell CMake that the project requires at least <min> but has been updated
  to work with policies introduced by <max> or earlier.

```
?

sedited replied:
Right.

fanquake replied:
Note that upstream have also done this now: https://github.com/google/leveldb/commit/9e273a6aef69ceb8717c083e46002641f6f51c8f.

davidgumberg replied:
I've pulled in the upstream change.

sedited reviewed (approved):
ACK bb48a453dae5d1faf972201fbd037e387add3e5a

fanquake commented on CMakeLists.txt:
```diff
@@ -11,8 +11,8 @@ set(CMAKE_C_STANDARD 11)
 
-# This project requires C++11.
-set(CMAKE_CXX_STANDARD 11)
+# This project requires C++17.
```
```suggestion
# This project requires C++20.
```

davidgumberg replied:
Marking as resolved since now the bump is to 17 instead of 20

pwnall committed:
Effectively, this change

* bumps the minimum CMake version to 3.22
* bumps the minimum C++ version to C++17

This aligns with the Google C++ support policy at
https://github.com/google/oss-policies-info/blob/main/foundational-cxx-support-matrix.md

PiperOrigin-RevId: 881717521

fanquake commented:
@davidgumberg could you circle back here?

davidgumberg commented:
Pushed to instead cherry pick the cmake version and cxx version bumps from upstream (https://github.com/bitcoin-core/leveldb-subtree/pull/55/commits/d123cf5a8396074e5d9cbd860b618eb39789df6b).

Building works if benchmark building is disabled now:

```bash
cmake -B build -DLEVELDB_BUILD_BENCHMARKS=off 
cmake --build build
```

hebasto reviewed (approved):
ACK d123cf5a8396074e5d9cbd860b618eb39789df6b.
